Signal/src/call/PeerConnectionClient.swift

```diff
@@ -74,7 +74,31 @@ protocol PeerConnectionClientDelegate: class {
  */
 class PeerConnectionClient: NSObject, RTCPeerConnectionDelegate, RTCDataChannelDelegate {
 
-    let TAG = "[PeerConnectionClient]"
+    private class AtomicHandle<ValueType> : NSObject {
```

Comment:
In Swift, weak variables aren't thread safe.  It isn't safe to resolve/acquire/lock a weak reference into a strong reference while the instance might be being deallocated on another thread.  

`AtomicHandle` provides thread-safe access to a strong reference.  `PeerConnectionClient` has an `AtomicHandle` to itself that its many async blocks (which run on more than one thread) can use to safely try to acquire a strong reference to the `PeerConnectionClient`.  In ARC we'd normally, we'd avoid having an instance retain a strong reference to itself to avoid retain cycles, but it's safe in this case: `PeerConnectionClient` is owned (and only used by) a single entity `CallService` and `CallService` always calls `[PeerConnectionClient terminate]` when it is done with a `PeerConnectionClient` instance, so `terminate` is a reliable place where we can break the retain cycle.

Reply:
Note: this fixes the symptom of `EXC_BAD_ACCESS` during `PeerConnectionClient` teardown which I'd seen for a long time but misattributed to WebRTC because WebRTC entities like `RTCVideoTrack` cause similar errors if they are not deallocated in a timely when during call teardown.

Reply:
I added a lengthy comment.